lib/learn_to_rank/data_pipeline/embed_features.rb

```diff
@@ -48,20 +47,30 @@ def features(judgement, cached_queries)
       { features: feats }
     end
 
-    def fetch_document(judgement, cached_queries)
+    def fetch_document(judgement)
+      unless judgement[:query] == @last_query
+        @last_query = judgement[:query]
+        @last_results = do_fetch(judgement)
+      end
+
+      unless @last_results.nil?
+        @last_results.find { |doc| doc["link"] == judgement[:link] }
```

Comment:
A mini optimization here would be to create a new hash for the results indexed by link, so you don't need to do a find.